File: hacking_with_php/CSTI.php
<?php
// Function to send HTTP requests
function sendRequest($url, $method = 'GET', $data = null) {
    $options = array(
        'http' => array(
            'method' => $method,
            'ignore_errors' => true // Ignore HTTP errors to capture the response
        )
    );

    if ($data !== null) {
        $options['http']['content'] = http_build_query($data);
    }

    $context = stream_context_create($options);
    return file_get_contents($url, false, $context);
}

// Function to check for CSTI vulnerability
function testCSTIVulnerability($url) {
    $response = sendRequest($url);
    if (strpos($response, '{{') !== false && strpos($response, '}}') !== false) {
        echo "\033[0;32mThe target website may be vulnerable to Client-Side Template Injection!\033[0m\n\n";
        echo "Here's how to test the vulnerability:\n";
        echo "1. Inject a simple expression like {{1+1}} in input fields or URL parameters.\n";
        echo "2. If the website evaluates the expression and displays the result (e.g., 2), it's likely vulnerable.\n";
        echo "3. Perform further testing by injecting more complex expressions to confirm the extent of vulnerability.\n";
    } else {
        echo "\033[0;31mThe target website does not appear to be vulnerable to Client-Side Template Injection.\033[0m\n";
    }
}

// Main program
echo "Enter your target website URL: ";
$targetWebsite = trim(fgets(STDIN));

echo "Testing for CSTI vulnerability on $targetWebsite...\n\n";

// Performing the vulnerability test
testCSTIVulnerability($targetWebsite);

File: hacking_with_perl/CSTI.pl
#!/usr/bin/perl

use strict;
use warnings;
use LWP::UserAgent;
use Term::ANSIColor;

# Function to check if the given URL is vulnerable to CSTI
sub check_vulnerability {
    my ($url) = @_;
    my $ua = LWP::UserAgent->new;
    my $response = $ua->get($url);
    my $content = $response->decoded_content;
    
    if ($content =~ /{{.*?}}/) {
        return 1; # Vulnerable
    } else {
        return 0; # Not vulnerable
    }
}

# Function to display result with color
sub display_result {
    my ($url, $vulnerable) = @_;
    print "Target website: $url\n";
    if ($vulnerable) {
        print color('red'), "Vulnerable to Client-Side Template Injection!\n", color('reset');
        print "You can test the vulnerability by injecting {{alert('XSS')}} into input fields.\n";
    } else {
        print color('green'), "Not vulnerable to Client-Side Template Injection.\n", color('reset');
    }
}

# Main program
print "Enter your target website URL: ";
my $target_url = <STDIN>;
chomp $target_url;

print "Checking for vulnerability...\n";
my $is_vulnerable = check_vulnerability($target_url);

# Display result
display_result($target_url, $is_vulnerable);
